src/common/com/intellij/plugins/haxe/lang/psi/impl/AbstractHaxePsiClass.java

```diff
@@ -72,15 +72,29 @@ public HaxeNamedComponent getTypeComponent() {
   @NotNull
   @Override
   public String getQualifiedName() {
-    final String name = getName();
+    String name = getName();
     if (getParent() == null) {
       return name == null ? "" : name;
     }
+
+    if(name == null && this instanceof HaxeAnonymousType) {
+      // restore name from parent
+      HaxeTypedefDeclaration typedefDecl = (HaxeTypedefDeclaration)(getParent().getParent());
+      if(typedefDecl != null) {
+        name = typedefDecl.getName();
+      }
+    }
+
     final String fileName = FileUtil.getNameWithoutExtension(getContainingFile().getName());
     String packageName = HaxeResolveUtil.getPackageName(getContainingFile());
-    if (isAncillaryClass(name, fileName)) {
+
+    // Check type is `Ancillary` if:
+    // 1. NOT root package! (packageName is NOT empty)
+    // 2. `name` is not null
+    if (name != null && !packageName.isEmpty() && isAncillaryClass(name, fileName)) {
```

Comment:
Yes, it's made for root packages:
- Types.hx declares Bar1 and Bar2, QNames for them - Bar1, Bar2
- tests/Types2.hx declares Foo1 and Foo2, QNames for them - tests.Types2.Foo1, tests.Types2.Foo2

http://haxe.org/manual/type-system-modules-and-paths.html - and sub topics
For root package - names accessable from anywhere, so QName is just name of type (look at StdTypes.hx which is globally provide basic types)